### PR TITLE
Reorganize project and add new av-cod-server service with start/stop

### LINKED ISSUE #30: Implement av-cod-server for start/end
Implement initial prototype `av-cod-server` process which implements design from #28 (PR #29)

### PR DESCRIPTION
**Pull request by ajfabbri**
Goal is to implement v0 of client-side (autonomous vehicle, or "AV" side) API for Trial Start and Trial End messages. Being a separate process, intended to run in a container, the API is exposed via HTTP REST.

- Reorganize stuff into subdirectories now that there are three prototypes brewing in this repo.
- Initial implementation of `av-cod-server` with support for start / end APIs.

Closes #30 

**Update:** This is ready for review. I should probably squash the last two commits together--but felt like sharing the pain and lessons learned from using a 3rd party http server dependency, to ditching that and going direct with the lower-level node http APIs.

### CODE REVIEW

File: src/util/cond_promise.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+            this.resolveCb = resolve
+            this.rejectCb = reject
+        })
+    }
+    getPromise() {
+        return this.promise
+    }
+    resolve() {
+        setImmediate(() => this.resolveCb())
+    }
```
Comment (ajfabbri) [inline src/util/cond_promise.ts:23]: I don't know if `setImmediate()` is required here. The story is, node has this thing [EventEmitter](https://nodejs.org/en/learn/asynchronous-work/the-nodejs-event-emitter) which is used a lot--including for events that fire from the node `http.Server` class we use. Those `on()` callbacks are called synchronously--so to be cautious I don't call `this.resolveCb()` directly from `resolve()` (which is called via an `on()` event handler). Instead, I tell the node async runtime to call the `resolveCb()` on the next tick of its event loop. This is what happens when you're experienced with async code but a noob at javascript / node. 😅

File: src/av-cod-server/http_server.ts [hunk 2]
```diff
@@ -9,79 +10,111 @@ import {
...
-        start()
+        this.server.listen(options, () => {
+            console.info(`--> http server listening on ${options.host}:${port}`)
+        })
     }
 
     async stop() {
         console.info('<-- http server shutdown')
-        this.fastify.close()
+        this.server.close()
```
Comment (ajfabbri) [inline src/av-cod-server/http_server.ts:116]: Related to below comment: when you call `http.Server.close()`, it doesn't close immediately. It starts it async, and then fires an `Event` of `'close'`. So, to be able to `await` (block current task) until the server actually shuts down, you have to wait (put current task to sleep) until that `'close'` event is emitted. Thus I created this `CondPromise` thing above.

File: src/av-cod-server/http_server.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        const id = new TrialId()
+        rep.writeHead(200, { 'Content-Type': JSON_CONTENT })
+        rep.end(new v0TrialEnd(ts, id).serialize())
+    }
+
+    private async registerRoutes() {
+        this.server.on(
+            'request',
+            (req: IncomingMessage, res: ServerResponse) => {
+                if (req.method !== 'GET') {
```
Comment (kitplummer) [inline src/av-cod-server/http_server.ts:60]: Pretty sure this should be a POST.

File: src/av-cod-server/http_server.test.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
+import http, { IncomingMessage } from 'node:http'
+import { HttpServer } from './http_server.js'
+import { Config } from './config.js'
+import { DittoCOD } from '../ditto_cod.js'
+import { sleep } from '../util/util.js'
+
+// convenience wrapper around node's low-level http interface
+async function httpGet(url: string): Promise<string> {
+    return new Promise((resolve, reject) => {
+        http.get(url, (res: IncomingMessage) => {
```
Comment (kitplummer) [inline src/av-cod-server/http_server.test.ts:10]: Commented in impl that I believe this should be a POST.

**Review by kitplummer — CHANGES_REQUESTED** [review]
In general, I believe the practice is when a `data` block is passed with the HTTP request it needs to be either PUT or POST.  In REST-y circles - that usually also addresses CREATE or UPDATE actions.  But most modern HTTP clients won't require extra headers to handle data if it is a POST (ala cURL).  I know the contact reports are POST'd - as defined in the MAPG ICD - but didn't look to see if anything else is specifically POST or GET.  So think we should do what is the norm here.

**Comment by ajfabbri** [comment]
I am thinking
GET status/start/stop I.e. "give me status please"
PUT contact reports, telemetry I.e. "I'm creating a new contact report, etc."
Telemetry could maybe be POST.. I'm not there yet.

IIUC, you're saying PUT/POST w/ an empty request to query status? 

Aside, for your entertainment: I actually think the semantic binding with http verbs and general purpose APIs is overhyped (a layering violation except for trivial CRUD APIs). Personally I'm cool with "always POST" but people seem not to like it 😅.

**Comment by ajfabbri** [comment]
Oh I see your update now. Thanks for the review. 😅

**Pull request merged**